api/src/main/java/org/eclipse/microprofile/config/spi/ConfigValue.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.eclipse.microprofile.config.spi;
+
+import java.util.Objects;
+
+/**
+ * The ConfigValue is a metadata object that holds additional information after the lookup of a configuration.
+ * <p>
+ *
+ * Right now, it is able to hold information like the configuration name, value, the Config Source from where
+ * the configuration was loaded, the ordinal of the Config Source and a line number from where the configuration was
+ * read if exists.
+ * <p>
+ *
+ * This is used together with {@link ConfigValueConfigSource} and {@link ConfigSourceInterceptor} to expose the
+ * Configuration lookup metadata.
+ *
+ * @author <a href="mailto:radcortez@yahoo.com">Roberto Cortez</a>
+ */
+public class ConfigValue {
+    private final String name;
+    private final String value;
+    private final String configSourceName;
+    private final int configSourceOrdinal;
+
+    private final int lineNumber;
+
+    private ConfigValue(final ConfigValueBuilder builder) {
+        this.name = builder.name;
+        this.value = builder.value;
+        this.configSourceName = builder.configSourceName;
+        this.configSourceOrdinal = builder.configSourceOrdinal;
+        this.lineNumber = builder.lineNumber;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public String getValue() {
+        return value;
+    }
+
+    public String getConfigSourceName() {
+        return configSourceName;
+    }
+
+    public int getConfigSourceOrdinal() {
+        return configSourceOrdinal;
+    }
+
+    public int getLineNumber() {
+        return lineNumber;
+    }
```

Comment:
I think we should consider an abstraction for source info. Not all sources will be line-oriented.

Reply:
The problem with this approach is that every abstraction requires another instance to allocate, and that can add up quickly.  See also #552.

Reply:
If we're going to prioritize performance, then we should be certain there's a clear need for this sort of traceability. Do we have a use case for this one?

Reply:
Yes.  In Quarkus we want to be able to report the source and the line number when a configuration validation error occurs.

Reply:
Well at least we wanted to report what is the source that load a certain Config, so having the ConfigSource name is useful.

Priority could help you make decisions when you are intercepting stuff.

LineNumber works for file bases sources, but can work for databases as well. Maybe that can be renamed to a more neutral name.

Reply:
I can see the motivation, I'm just curious if this is something that users are asking for. I definitely like the idea of identifying the config source that provided the value, but line number is obviously only useful in strictly line-oriented files - so essentially just `.properties` files in our ecosystem - and once you know the file and the property name, the  line number is not terribly helpful for locating the value.

I'm not opposed to including the data point if it's useful for someone, but it seems like, if that is the case, we ought to make it useful in other cases too. For instance, in a YAML or JSON file you might want the column number in addition to the line number. Also, I suppose including the line number presumes that the file name is somehow included in the config source name, but that's not really clear.

As to the object allocation concern, it seems like sacrificing flexibility for performance here is a departure from other design decisions in the API. Can we come up with an optional abstraction and leave implementations free to include their own optimized data points under the covers?

Maybe instead of `int getLineNumber()` we could have `String getSourceLocation()`. Then a source could return `/foo/bar/file.properties (line: 42)` or `/foo/bar/file.json (line: 42, col: 13)` or `jdbc:mysql://foo:3306/bar (table: properties)`, and we can default to `n/a`.

Reply:
It doesn't integrate well when you have a diagnostics infrastructure or tooling integration of some sort and you need the line number as a piece of data.  Another problem is that you either have to create a new string on every config property lookup (even if the value isn't used) or else you have to keep another string for every config key.  Keeping an int costs almost nothing.

Reply:
Sure, but if we entertain tooling or diagnostic use cases that need specific information, that's all the more reason to provide a better abstract that can support a richer set of config sources.

If the actual use case is "report the source and the line number when a configuration validation error occurs", then a string suffices.

My main concern here is that if we include `getLineNumber()` in the API, we open the door for users to expect similar context for other config sources without an abstraction to accommodate it.

Maybe we change ConfigValue to an interface (without `getLineNumber()`) and let implementations extend as they see fit.

We could even specify mix-in interfaces with common functionality that implementations could standardize on.

e.g.
```java
public interface LineOriented {
    int getLineNumber();
}
```
```java
public interface FileBased {
    Path getPath();
}
```

Reply:
The trouble with mixins is that every config source and interceptor which must construct an instance has to agree to construct the same type; interceptors in particular would have a hard time figuring this out.  A location class is at least better in this regard.

Reply:
Oh, right, I overlooked the the part where interceptor implementors need to construct the value containers. Forcing interceptors to propagate hidden interfaces is certainly not ideal :P Of course, if we're relying on every participant to contribute appropriate tracing information, that's not a very reliable solution anyway.

I'll rest my case here since this thread is getting lengthy for a PR comment. Including lineNumber is not a showstopper for me. Sounds like we're aware of the risks and benefits.

Reply:
Well, a location type abstraction is not a showstopper for me either, I guess.  In the worst case it seems like interceptors would be propagating them verbatim from the config source, and hopefully the config source would not have to instantiate them ad-hoc (or if it does, it would be able to cache them) in cases where the config source would otherwise be static (i.e. efficient).